People/davidgamba/rmd 20676 refactor beginscan task

DavidGamba opened the pull request:
@tonytvo please review.

Commit:
rmd-20676 split tests into different classes

Commit:
rmd-20676 Tidy up UploadFile tests

Commit:
rmd-20676 Refactor GetPreScanResults task

* Compile statically.
* Add tests.

Commit:
rmd-20676 Refactor BeginScan task

* Add moduleWhitelist field to VeracodeSetup extension.
  - modules are provided as a String Set so the user is free to obtain
  that Set from files or however they want instead of being forced to
  have it in a file as before.
* Compile statically.
* Add tests.

Commit:
rmd-20676 Load VeracodeAPI only once

The task constructor gets evaluated several times during a run.
Load the API only once and fail if it is missing credentials.

Commit:
rmd-20676 Load veracodeSetup when determining InputFiles

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -26,28 +26,78 @@
+        Set<String> moduleIds = []
+        Set<String> moduleNames = []
+        NodeList moduleList = xml.getAt("module") as NodeList
+        for (int i = 0; i < moduleList.size(); i++) {
```
sounds like this loop is to extract any moduleIds that has status doesn't start with Fatal? then should we extract it to something? extractModulesDoesNotHavePrevisFatal?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -26,28 +26,78 @@
-        println "Modules selected: ${moduleIds.size()}"
-        if (whiteList.size() != moduleIds.size()) {
-            println 'WARNING: Not all the files in whitelist are being scanned. Some modules no longer exist? Manual whitelist maintenance should be performed.'
+        if (whitelist.size() != moduleNames.size()) {
```
what does this if statement do?

DavidGamba replied:
Checks if the whitelist is bigger than the whitelist modules found. If that is the case then it "warns" that not all whitelist modules where found.

tonytvo replied:
I see ... so you might want to extract this to something like warnIfNotAllWhitelistModulesFound?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeAPI.groovy:
```diff
@@ -44,15 +45,29 @@ class VeracodeAPI {
 
     private boolean useAPICredentials() {
-        if (this.username != "" && this.password != "") {
+        if (this.username != null && this.password != null &&
```
sounds like you can use groovy truth here .... null and empty string will return false .... so you can just use if (!username and !password) is probably good enough

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeAPI.groovy:
```diff
@@ -44,15 +45,29 @@ class VeracodeAPI {
+                this.username != "" && this.password != "") {
             return false
         }
+        if (this.key == null && this.id == null) {
```
similarly you could use if (!this.key && !this.id)...

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -26,28 +26,78 @@
+    void run() {
+        this.moduleWhitelist = veracodeSetup.moduleWhitelist
+        Set<String> moduleIds = extractWhitelistModuleIds(readXml(preScanResultsOutputFile), moduleWhitelist)
+        Node xml = writeXml(
```
do we need to write both to outputfile and printf to console line at the same time? can we just use log4j to print it out to console first and we could configure it to print to output file if need to?

DavidGamba replied:
Writting to file allows to cache the output in the tasks where it makes sense and so gradle doesn't have to actually run the task again if the `@outputFile` is there.

The content written to the console is only the relevant subset of the full xml.

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeTask.groovy:
```diff
@@ -124,8 +132,12 @@ abstract class VeracodeTask extends DefaultTask {
-    protected def readXml(String filename) {
-        new XmlParser().parseText(GFileUtils.readFile(new File(filename)))
+    protected Node readXml(File file) {
+        new XmlParser().parseText(GFileUtils.readFile(file))
```
I think groovy has parse(file) .... so you can use new XmlParser().parse(file)

tonytvo commented on src/test/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTest.groovy:
```diff
@@ -0,0 +1,80 @@
+
+    def 'Test printBeginScanStatus'() {
+        given:
+        def os = mockSystemOut()
```
I didn't know you could use mockSystemOut()... that's cool

tonytvo commented on src/test/groovy/com/calgaryscientific/gradle/VeracodeGetPreScanResultsTest.groovy:
```diff
@@ -0,0 +1,68 @@
+package com.calgaryscientific.gradle
+
+class VeracodeGetPreScanResultsTest extends VeracodeTaskTest {
+    static String successXMLResponse = '''
```
it seems like you could extract these string to xml files in the test/resources folder .... and give it a proper name....

DavidGamba replied:
Like all the XML responses from Veracode in a single file?

tonytvo commented on src/test/groovy/com/calgaryscientific/gradle/VeracodeTaskTest.groovy:
```diff
@@ -72,6 +72,11 @@ class VeracodeTaskTest extends Specification {
         System.out = stdout
     }
 
+    Node parseXMLString(String xml) {
```
hm... it seems like we have lots of XMLIO in multiple places (veracodeTask, and now this) ... should we tract all of the XMLIO behavior to its own class? like XMLManipulation?

DavidGamba replied:
Sounds good, but lets get that in another PR, this one is already getting big.

tonytvo commented on src/test/groovy/com/calgaryscientific/gradle/VeracodeUploadFileTest.groovy:
```diff
@@ -0,0 +1,106 @@
+
+class VeracodeUploadFileTest extends VeracodeTaskTest {
+
+    String successXMLResponse = '''<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
similarly here, we could extract this to xml file in test/resources ... and see if we could reduce any duplication

tonytvo commented on src/test/groovy/com/calgaryscientific/gradle/VeracodeUploadFileTest.groovy:
```diff
@@ -0,0 +1,106 @@
+        task.run()
+
+        then:
+        1 * task.veracodeAPI.uploadFile(_, _) >> {
```
would it be beneficial, if some how we could capture the response of uploadfile and inspect the response of the uploadfile? with this ... we know that the uploadfile has been executed and we expect the response is in certain required format?

DavidGamba replied:
This test verifies that the printFileUploadStatus method can properly parse the response XML:
https://github.com/calgaryscientific/veracode-gradle-plugin/pull/15/files/2428ce943c8a3e26970554f8e3e55c5e542fa88a#diff-9e7257a8f5f326082d82894db9c4d331R93

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -26,28 +26,78 @@
-        writeXml('build/scan.xml', uploadAPI().beginScan(project.app_id, moduleIds.join(","), 'false'))
+    void run() {
+        this.moduleWhitelist = veracodeSetup.moduleWhitelist
+        Set<String> moduleIds = extractWhitelistModuleIds(readXml(preScanResultsOutputFile), moduleWhitelist)
```
would it be easier for us to test this function, if we pass in the string parameter, instead of Node parameter (and the string parameter could be the content of outputfile or the filepath name to the file? that way we don't have to have some setup code to parse the content of the file into Node.....

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeBeginScanTask.groovy:
```diff
@@ -26,28 +26,78 @@
+    }
 
-        writeXml('build/scan.xml', uploadAPI().beginScan(project.app_id, moduleIds.join(","), 'false'))
+    void run() {
```
what does this run function do? and do we have test to document that behavior?

tonytvo commented on src/main/groovy/com/calgaryscientific/gradle/VeracodeGetPreScanResultsTask.groovy:
```diff
@@ -26,24 +26,50 @@
-        String xmlResponse
+        String response
+        String file
         if (project.hasProperty('build_id')) {
```
what's difference between build_id and app_id?

DavidGamba replied:
The app is the product itself, resmd in our case, you could be scanning for multiple products. The build_id is the id of the actual veracode scan request for that app.

Commit:
rmd-20676 Use groovy truth to ensure credentials exist

Commit:
rmd-20676 Ensure BeginScan task uses a given build_id

* Before, the task was relying on the latest pre scan results.

Commit:
rmd-20676 Split extractWhitelistModuleIds function into smaller components

Commit:
rmd-20676 Add beginScan method to VeracodeAPI

Commit:
rmd-20676 use XmlParser().parse to read a file

DavidGamba commented:
@tonytvo I think I addressed all of the comments with the exception of 2 that I would prefer in another PR:
* Move XML handling code to a single class.
* Move XML responses to a single resource file.
